flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/BlockStatementSplitter.java

```diff
@@ -0,0 +1,331 @@
+ *     } else {
+ *         b = a * 3;
+ *         System.out.println(b);
+ *     }
+ *     counter--;
+ * }
+ *
+ * </code></pre>
+ *
+ * <p><i>After</i>
+ *
+ * <pre><code>
+ * while (counter > 0) {
+ *     myFun_whileBody0_0(int a);
+ *     if (a > 0) {
+ *         myFun_whileBody0_0_ifBody0(int a);
+ *     } else {
+ *         myFun_whileBody0_0_ifBody1(int a);
+ *     }
+ *     counter--;
+ * }
+ * </code></pre>
+ *
+ * <p>Where bodies of extracted "methods" are:
+ *
+ * <pre><code>
+ * myFun_whileBody0_0(int a) ->
+ *     int localA = a + 1000;
+ *     System.out.println(localA);
+ * </code></pre>
+ *
+ * <pre><code>
+ * myFun_whileBody0_0_ifBody0(int a) ->
+ *     b = a * 2;
+ *     c = b * 2;
+ *     System.out.println(b);
+ * </code></pre>
+ *
+ * <pre><code>
+ * myFun_whileBody0_0_ifBody1(int a) ->
+ *     b = a * 3;
+ *     System.out.println(b);
+ * </code></pre>
+ */
+@Internal
+public class BlockStatementSplitter {
+
+    private final String code;
+
+    private final String parameters;
+
+    private BlockStatementVisitor visitor;
+
+    /**
+     * Initialize new BlockStatementSplitter.
+     *
+     * @param code a code block that should be rewritten.
+     * @param parameters parameters definition that should be used for extracted methods.
+     */
+    public BlockStatementSplitter(String code, String parameters) {
+        this.code = code;
+        this.parameters = parameters;
+    }
+
+    /**
+     * This method extracts statements from IFs, ELSE's and WHILE blocks from block code used during
+     * initialization of this object. Every entry of returned map can be seen as new method nam (map
+     * key) and method's body. The block names will be prefixed with provided context.
+     *
+     * @param context prefix for extracted blocks.
+     * @return a map of block name to block statements mappings. The key can be interpreted as name
+     *     of extracted block/method and corresponding List represents individual statements (block'
+     *     lines) for this block.
+     */
+    public Map<String, List<String>> extractBlocks(String context) {
+
+        this.visitor = new BlockStatementVisitor(code, context, parameters);
+        JavaParser javaParser = new JavaParser(visitor.tokenStream);
+        javaParser.getInterpreter().setPredictionMode(PredictionMode.SLL);
+        visitor.visitStatement(javaParser.statement());
+        return visitor.getAllBlocks();
+    }
+
+    /**
+     * Rewrite code block that was used for this object initialization.
+     *
+     * @return a map which key represent rewritten block name and value represents rewritten code
+     *     block, including calls to extracted methods
+     */
+    public Map<String, String> rewriteBlock() {
+        visitor.rewrite();
+        Map<String, String> rewriteBlocks = new HashMap<>();
+        rewriteBlocks.put(visitor.context, visitor.rewriter.getText());
+        return rewriteBlocks;
+    }
+
+    private static void mergeBlocks(
```

Comment:
Why do you need to `mergeBlocks`? Statements from different levels of blocks may be extracted into the same new method?

Reply:
The level of block is identified by its own "context" value. So The context value become automatically the name of extracted method.
Blocks with the same "context" should be extracted to the same method.

In here I could use `map.putAll` but then there would be a risk to override values for already existing key (context).

Reply:
maybe `mergeBlocks` is not the best name in this case.